[@property] Support <transform-function> and <transform-list> types

anttijk opened the pull request:
#### 41d246922c5892281b203d3deb209bac0501a13f
<pre>
[@property] Support &lt;transform-function&gt; and &lt;transform-list&gt; types
<a href="https://bugs.webkit.org/show_bug.cgi?id=249465">https://bugs.webkit.org/show_bug.cgi?id=249465</a>
rdar://103441980

Reviewed by Antoine Quint.

Add the remaining syntax types.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-initial-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/self-utils-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::transformOperationAsCSSValue):

Factor this into a function so we can use it to serialize TransformOperation values in CSSCustomPropertyValue.

(WebCore::computedTransform):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
(WebCore::CSSCustomPropertySyntax::parseComponent):
(WebCore::CSSCustomPropertySyntax::typeForTypeName):

Also use SortedArrayMap.

* Source/WebCore/css/parser/CSSCustomPropertySyntax.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeCustomPropertyValueWithSyntax):
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTransformFunction):
(WebCore::CSSPropertyParserHelpers::consumeTransform):
(WebCore::CSSPropertyParserHelpers::consumeTransformValue): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebKit/WebKit.xcodeproj/xcshareddata/xcschemes/WebKit.xcscheme:

Canonical link: <a href="https://commits.webkit.org/257987@main">https://commits.webkit.org/257987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d553e55c89f8b76d628725ee267871b406f3092

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100607 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33646 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109910 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170187 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104591 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10683 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93004 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107759 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106387 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3460 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24246 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3474 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43740 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5267 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/css/CSSCustomPropertyValue.h:
```diff
@@ -45,8 +46,8 @@ class CSSCustomPropertyValue final : public CSSValue {
 
         bool operator==(const NumericSyntaxValue& other) const { return value == other.value && unitType == other.unitType; }
     };
-    
```
Nice.

graouts commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -363,6 +364,10 @@ std::pair<RefPtr<CSSValue>, CSSCustomPropertySyntax::Type> CSSPropertyParser::co
+        case CSSCustomPropertySyntax::Type::TransformFunction:
+            return consumeTransformFunction(m_range, m_context);
+        case CSSCustomPropertySyntax::Type::TransformList:
+            return consumeTransform(m_range, m_context);
```
Would be nice to change this to be called `consumeTransformList()` since it's no longer specific to the `transform` property.

graouts commented on Source/WebKit/WebKit.xcodeproj/xcshareddata/xcschemes/WebKit.xcscheme:
```diff
@@ -31,7 +31,7 @@
    </TestAction>
    <LaunchAction
-      buildConfiguration = "Release"
+      buildConfiguration = "Debug"
```
Did you intend to have this in your patch?

graouts commented on Source/WebKit/WebKit.xcodeproj/xcshareddata/xcschemes/WebKit.xcscheme:
```diff
@@ -61,7 +61,7 @@
    </AnalyzeAction>
    <ArchiveAction
-      buildConfiguration = "Release"
+      buildConfiguration = "Debug"
```
Probably want to get rid of this as well.

graouts commented on Source/WebCore/css/CSSCustomPropertyValue.cpp:
```diff
@@ -79,6 +79,11 @@ String CSSCustomPropertyValue::customCSSText() const
         }, [&](const String& value) {
             return value;
+        }, [&](const RefPtr<TransformOperation>& value) {
+            auto cssValue = transformOperationAsCSSValue(*value, RenderStyle::defaultStyle());
+            if (!cssValue)
+                return emptyString();
+            return cssValue->cssText();
```
Line-saving opportunity!

```suggestion
            if (auto cssValue = transformOperationAsCSSValue(*value, RenderStyle::defaultStyle()))
                return cssValue->cssText();
            return emptyString();
```

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -732,6 +732,152 @@ static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& tr
+    };
+
+    switch (operation.type()) {
+        // translate
```
Misaligned.

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -732,6 +732,152 @@ static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& tr
+            functionValue->append(translateLengthAsCSSValue(translate.z()));
+        return functionValue;
+    }
+        // scale
```
Misaligned.

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -732,6 +732,152 @@ static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& tr
+            functionValue->append(cssValuePool.createValue(scale.z(), CSSUnitType::CSS_NUMBER));
+        return functionValue;
+    }
+        // rotate
```
Misaligned.

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -732,6 +732,152 @@ static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& tr
+        functionValue->append(cssValuePool.createValue(rotate.angle(), CSSUnitType::CSS_DEG));
+        return functionValue;
+    }
+        // skew
```
Misaligned.

anttijk committed:
[@property] Support <transform-function> and <transform-list> types
https://bugs.webkit.org/show_bug.cgi?id=249465
rdar://103441980

Reviewed by Antoine Quint.

Add the remaining syntax types.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-initial-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/self-utils-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::transformOperationAsCSSValue):

Factor this into a function so we can use it to serialize TransformOperation values in CSSCustomPropertyValue.

(WebCore::computedTransform):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
(WebCore::CSSCustomPropertySyntax::parseComponent):
(WebCore::CSSCustomPropertySyntax::typeForTypeName):

Also use SortedArrayMap.

* Source/WebCore/css/parser/CSSCustomPropertySyntax.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeCustomPropertyValueWithSyntax):
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTransformFunction):
(WebCore::CSSPropertyParserHelpers::consumeTransform):
(WebCore::CSSPropertyParserHelpers::consumeTransformValue): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebKit/WebKit.xcodeproj/xcshareddata/xcschemes/WebKit.xcscheme:

Canonical link: https://commits.webkit.org/257987@main
